products/libllbuild/public-api/llbuild/buildsystem.h

```diff
@@ -319,6 +319,15 @@ typedef struct llb_buildsystem_delegate_t_ {
                                    llb_buildsystem_process_t* process,
                                    llb_buildsystem_command_result_t result,
                                    int exit_status);
+
+  /// Called when a cycle is detected by the build engine and it cannot make
+  /// forward progress.
+  ///
+  /// Xparam rules The ordered list of items comprising the cycle, starting from
+  /// the node which was requested to build and ending with the first node in
+  /// the cycle (i.e., the node participating in the cycle will appear twice).
+  /// Xparam ruleCount The number of rules comprising the cycle.
+  void (*cycleDetected)(void* context, llb_data_t* rules, uint64_t ruleCount);
```

Comment:
s/cycleDetected/cycle_detected/, s/ruleCount/rule_count/ -- the C API uses traditional C / POSIX style naming.